api/v1alpha1/serviceresolver_types.go

```diff
@@ -0,0 +1,299 @@
+	// expression can filter on the same selectors as the Health API endpoint.
+	Filter string `json:"filter,omitempty"`
+	// OnlyPassing specifies the behavior of the resolver's health check
+	// interpretation. If this is set to false, instances with checks in the
+	// passing as well as the warning states will be considered healthy. If this
+	// is set to true, only instances with checks in the passing state will be
+	// considered healthy.
+	OnlyPassing bool `json:"onlyPassing,omitempty"`
+}
+
+type ServiceResolverFailover struct {
+	// Service is the service to resolve instead of the default as the failover
+	// group of instances during failover.
+	Service string `json:"service,omitempty"`
+	// ServiceSubset is the named subset of the requested service to resolve as
+	// the failover group of instances. If empty the default subset for the
+	// requested service is used.
+	ServiceSubset string `json:"serviceSubset,omitempty"`
+	// Namespace is the namespace to resolve the requested service from to form
+	// the failover group of instances. If empty the current namespace is used.
+	Namespace string `json:"namespaces,omitempty"`
+	// Datacenters is a fixed list of datacenters to try during failover.
+	Datacenters []string `json:"datacenters,omitempty"`
+}
+
+func init() {
+	SchemeBuilder.Register(&ServiceResolver{}, &ServiceResolverList{})
+}
+
+func (in ServiceResolverSubsetMap) toConsul() map[string]capi.ServiceResolverSubset {
+	if in == nil {
+		return nil
+	}
+	m := make(map[string]capi.ServiceResolverSubset)
+	for k, v := range in {
+		m[k] = v.toConsul()
+	}
+	return m
+}
+
+func (in ServiceResolverSubsetMap) matchesConsul(candidate map[string]capi.ServiceResolverSubset) bool {
+	if len(in) != len(candidate) {
+		return false
+	}
+
+	for thisKey, thisVal := range in {
+		candidateVal, ok := candidate[thisKey]
+		if !ok {
+			return false
+		}
+		if !thisVal.matchesConsul(candidateVal) {
+			return false
+		}
+	}
+	return true
+}
+
+func (in ServiceResolverSubset) toConsul() capi.ServiceResolverSubset {
+	return capi.ServiceResolverSubset{
+		Filter:      in.Filter,
+		OnlyPassing: in.OnlyPassing,
+	}
+}
+
+func (in ServiceResolverSubset) matchesConsul(candidate capi.ServiceResolverSubset) bool {
+	return in.OnlyPassing == candidate.OnlyPassing && in.Filter == candidate.Filter
+}
+
+func (in *ServiceResolverRedirect) toConsul() *capi.ServiceResolverRedirect {
+	if in == nil {
+		return nil
+	}
+	return &capi.ServiceResolverRedirect{
+		Service:       in.Service,
+		ServiceSubset: in.ServiceSubset,
+		Namespace:     in.Namespace,
+		Datacenter:    in.Datacenter,
+	}
+}
+
+func (in *ServiceResolverRedirect) matchesConsul(candidate *capi.ServiceResolverRedirect) bool {
+	if in == nil || candidate == nil {
+		return in == nil && candidate == nil
+	}
+	return in.Service == candidate.Service &&
+		in.ServiceSubset == candidate.ServiceSubset &&
+		in.Namespace == candidate.Namespace &&
+		in.Datacenter == candidate.Datacenter
```

Comment:
```suggestion
		reflect.DeepEqual(in.Datacenters, candidate.Datacenters)
```

Reply:
This one is a string though I think. One has an array and the other has just a string.

Reply:
oohhhhhh...cool cool. that naming threw me off!